src/charm.py

```diff
@@ -4,101 +4,424 @@
+        basic_configurations = yaml.dump(
+            {
+                "log": {
+                    "level": "DEBUG",
+                },
+                "entryPoints": {
+                    "diagnostics": {
+                        "address": f":{self._diagnostics_port}"
+                    },
+                    "web": {
+                        "address": f":{self._port}"
+                    }
+                },
+                # We always start the Prometheus endpoint for simplicity
+                # TODO: Generate th
+                # is file in the dynamic configuration folder when the
+                # metrics-endpoint relation is established?
+                "metrics": {
+                    "prometheus": {
+                        "addRoutersLabels": True,
+                        "addServicesLabels": True,
+                        "entryPoint": "diagnostics",
+                    }
+                },
+                "ping": {
+                    "entryPoint": "diagnostics"
+                },
+                "providers": {
+                    "file": {
+                        "directory": _TRAEFIK_INGRESS_CONFIGURATIONS_DIRECTORY,
+                        "watch": True,
+                    }
+                }
+            }
+        )
+
+        container.push("/etc/traefik/traefik.yaml", basic_configurations)
+
+        # After the container (re)starts, we need to loop over the relations to ensure all
+        # the ingress configurations are there
+        for ingress_relation in self.model.relations["ingress"]:
+            self._process_ingress_relation(ingress_relation)
+
+        self._restart_traefik()
+
+        self.unit.status = ActiveStatus()
+
+    def _handle_ingress_change(self, event: RelationEvent):
+        if not isinstance(self.unit.status, ActiveStatus):
+            logger.debug("Charm not active yet, deferring event")
+            event.defer()
+            return
+
+        self._process_ingress_relation(event.relation)
+
+    def _process_ingress_relation(self, relation: Relation):
+        container = self.unit.get_container(_TRAEFIK_CONTAINER_NAME)
+
+        if not container.can_connect():
+            self.unit.status = WaitingStatus(
+                f"container '{_TRAEFIK_CONTAINER_NAME}' not yet ready"
+            )
+            return
+
+        other_app_name = relation.app.name
+
+        # Version negotiation
+
+        if self.unit.is_leader():
+            supported_versions_raw = relation.data[relation.app]["_supported_versions"]
+
+            supported_versions = yaml.safe_load(supported_versions_raw)
+
+            if "v3" not in supported_versions:
+                logger.error(
+                    f"The {other_app_name} application does not support the ingress relation v3 "
+                    f"(found: '{supported_versions}'); aborting data negotiation"
+                )
+                # TODO Put the charm in Blocked status?
+                return
```

Comment:
What happens with all these early returns? Feels like there should be some sensible return value resulting in a status being set?